### ISSUE TITLE
zone ls fails

### ISSUE THREAD

**Issue by arctica** [cockroachdb/cockroach#18617]
Running CockroachDB 1.0.6 on a very simple 3 node cluster.

```
# cockroach zone --insecure --host node1.example.com ls
Error: unexpected value: <nil>
Failed running "zone"
```

It first used to work. Then a new zone was defined for a table. That table was dropped and re-created. Now "zone ls" fails with the above error. I am able to set/get the zone infos without issue. Interestingly though, a "zone get" for the defined table zone showed a replica factor of 3 even though it was set to 1 before the table got dropped and re-created.

Somehow a NULL zone got entered into the system.zones table:

```
> select * from zones;
+----+----------------------------------------------------------------+
| id |                             config                             |
+----+----------------------------------------------------------------+
|  0 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
| 52 | NULL                                                           |
| 53 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x012\x00" |
+----+----------------------------------------------------------------+
(3 rows)
```

After deleting that row, "zone ls" works again.

**Comment by petermattis** [comment]
@arctica Thanks for the bug report. I wasn't able to reproduce the `NULL` entry in `system.zones`. Do you know exactly what steps you took?

> Interestingly though, a "zone get" for the defined table zone showed a replica factor of 3 even though it was set to 1 before the table got dropped and re-created.

This is expected, though not obvious. The zone is tied to the table's ID. When you drop and recreate a table the ID changes. Dropping a table or database should probably be deleting the associated zone.

@vivekmenezes We're likely not doing the correct thing on truncate now because truncate is creating a new table ID.

**Comment by arctica** [comment]
@petermattis thanks for the reply. I can reliably reproduce the NULL entry through the following steps:

1. Create table
2. Set replication factor for db.table specifically
3. Drop table
4. Create table

I suspect that due to the zone being tied to the table's ID, dropping the table results in a reference to a table that doesn't exist and somehow results in that NULL entry in system.zones. 

It might make sense to remove any replication zones that reference a table/db if said table/db get removed.

**Comment by petermattis** [comment]
@arctica Hmm, those are the exact steps I performed. I'll try this again on 1.0.6 (I'm running on master), perhaps something was fixed.

**Comment by petermattis** [comment]
> We're likely not doing the correct thing on truncate now because truncate is creating a new table ID.

Hmm, we seem to be doing this correctly. We also have code that deletes the zone config when a table is dropped. But the latter clearly isn't working correctly. I wonder why `TestDropTable` isn't catching this.

**Comment by arctica** [comment]
I just upgraded to v1.1-beta.20170907 and got this error:

```
# cockroach zone get test.t --insecure
Error: pq: column name "parentid" not found
Failed running "zone"
```

This is the content of system.zones:

```
> select * from system.zones;
+----+----------------------------------------------------------------+
| id |                             config                             |
+----+----------------------------------------------------------------+
|  0 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
| 54 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
+----+----------------------------------------------------------------+
(2 rows)
```

It seems this error is output for any zone that it couldn't find. It should probably have a more descriptive error like "zone not found". Not sure why it didn't find the zone though...

**Comment by petermattis** [comment]
@arctica Well that's not good. Can you reproduce from a fresh cluster and include the exact commands you're running (including the create and drop table commands)?

**Comment by arctica** [comment]
@petermattis I've started a completely fresh v1.0.6 cluster now and followed the following steps:

```
root@:26257/> create database test;
CREATE DATABASE
root@:26257/> select * from system.zones;
+----+----------------------------------------------------------------+
| id |                             config                             |
+----+----------------------------------------------------------------+
|  0 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
+----+----------------------------------------------------------------+
(1 row)
root@:26257/> CREATE TABLE test.foo (bar INT);
CREATE TABLE
```

```
# echo 'num_replicas: 1' | cockroach zone set test.foo -f - --insecure
range_min_bytes: 1048576
range_max_bytes: 67108864
gc:
  ttlseconds: 86400
num_replicas: 1
constraints: []
```

```
root@:26257/> SELECT * FROM system.zones;
+----+----------------------------------------------------------------+
| id |                             config                             |
+----+----------------------------------------------------------------+
|  0 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
| 51 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x012\x00" |
+----+----------------------------------------------------------------+
(2 rows)
root@:26257/> drop table test.foo;
DROP TABLE
root@:26257/> SELECT * FROM system.zones;
+----+----------------------------------------------------------------+
| id |                             config                             |
+----+----------------------------------------------------------------+
|  0 | "\x10\x80\x80@\x18\x80\x80\x80 \"\x04\b\x80\xa3\x05(\x032\x00" |
| 51 | NULL                                                           |
+----+----------------------------------------------------------------+
(2 rows)
```

**Comment by petermattis** [comment]
@arctica Thanks. The details are helpful.

**Comment by vivekmenezes** [comment]
I can confirm that the same behavior is happening with the 1.1 build. The only difference is that since DROP TABLE is implemented through an asynchronous code path in 1.1, the NULL entry shows up after a delay.

**Comment by vivekmenezes** [comment]
A related problem: If I use TRUNCATE instead of DROP I see this error:
```
root@:26257/test> TRUNCATE TABLE foo1;
pq: unable to marshal config.ZoneConfig: {[] 1048576 67108864 {86400} 1 {[]}}
```
